app/main.py

```diff
@@ -1,34 +1,107 @@
+            self.decks = [Deck(start[0], end[1])]
+        elif start[0] == end[0]:
+            self.decks = [Deck(start[0], i)
+                          for i in range(start[1], end[1] + 1)]
+        elif start[1] == end[1]:
+            self.decks = [Deck(i, start[1])
+                          for i in range(start[0], end[0] + 1)]
+
+    def get_deck(self, row: int, column: int) -> Deck:
+
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def fire(self, row: int, column: int) -> None:
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+        deck = self.get_deck(row, column)
+        deck.is_alive = False
+        if not list(filter(lambda deck: deck.is_alive, self.decks)):
+            self.is_drowned = True
+
+    def __repr__(self) -> str:
+        if self.is_drowned:
+            return "Drowned ship"
+        return "Alive ship"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+            self,
+            ships: list[tuple[tuple[int, int], tuple[int, int]]]
+    ) -> None:
+
+        self.ships = [Ship(*coordinates) for coordinates in ships]
+        self._validate_field()
+        self.field: dict = {}
+        for ship in self.ships:
+            ship_field = {(deck.row, deck.column): ship for deck in ship.decks}
+            self.field.update(ship_field)
+
+    def _validate_ships_count(self) -> None:
+
+        assert len(self.ships) == 10
+
+    def _validate_ship_sizes(self) -> None:
+
+        ship_sizes = {}
+        for ship in self.ships:
+            ship_sizes[len(ship.decks)] = ship_sizes.get(
+                len(ship.decks), 0
+            ) + 1
+        assert ship_sizes == {4: 1, 2: 3, 3: 2, 1: 4}
+
+    def _validate_field(self) -> None:
+
+        self._validate_ships_count()
+        self._validate_ship_sizes()
+
+    def fire(self, location: tuple[int, int]) -> str:
+
+        if location in self.field:
+            ship = self.field[location]
+            ship.fire(location[0], location[1])
+            if ship.is_drowned:
+                return "Sunk!"
+            else:
+                return "Hit!"
+        else:
+            return "Miss!"
+
+    def print_field(self) -> None:
+
+        for row in range(10):
```

Comment:
10 - what is the magic number? Move it to constant

Reply:
done